ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java

```diff
@@ -448,15 +461,15 @@ private boolean isFileBelowWatermark(FileStatus child, long highWatermark) {
     Path p = child.getPath();
     String fn = p.getName();
     if (!child.isDirectory()) {
-      return false;
+      return true;
     }
     if (fn.startsWith(AcidUtils.BASE_PREFIX)) {
       ParsedBaseLight b = ParsedBaseLight.parseBase(p);
-      return b.getWriteId() < highWatermark;
+      return b.getWriteId() <= highWatermark;
```

Comment:
This is good, but should also make sure that this base isn't the recently compacted base (AcidUtils#isCompactedBase should help).

Reply:
I don't think we would need any more checks - we already make sure that we ignore the files we use for the real acid dir:
* if we have anything beyond that below the watermark we should keep the request
* if there is nothing the request should be discarded...

I don't think we have any more problematic cases; do you have a testcase in mind?

Reply:
The AcidDirectory could be based off of some older txn id (depending on how the Cleaner selects which compaction to pick up – if CQ_NEXT_TXN_ID is filled in the cleaner will only attempt to clean when it can, however if CQ_NEXT_TXN_ID==NULL then the cleaner will try to clean even if it's blocked by an obsolete dir). It could be that txnId=99 was opened, table a was created, table a is compacted many times but the cleaner will never clean anything from table a because txid=99 sees table a as totally empty.

Reply:
I think this will not cause any problem - as I think this means the following:
* until txind=99 is closed the aciddir looks empty
* however we will have some dir which is not yet visible below the watermark => which will prevent the clean request from being removed

I don't think the above is incorrect behaviour - or could lead to the retention of `ready for cleaning` for indefinite times?

Reply:
You're right. Might be worth adding test cases to make sure that the Worker#isEnoughToCompact logic isn't changed though, because if it is then it will break the cleaner.
e.g. 
1. CQ_NEXT_TXN_ID is never updated (set it to NULL before the cleaner runs) and table dir contains base_7
2. CQ_NEXT_TXN_ID is not updated (set it to NULL before the cleaner runs) and table dir contains delta_1_7 and base_7